File: arcade/python/arcade-intro/07_Through the Fog/033_stringsRearrangement.py
def stringsRearrangement(inputArray):
    '''
    Given an array of equal-length strings, check if it is possible to rearrange the strings in 
    such a way that after the rearrangement the strings at consecutive positions would differ 
    by exactly one character.
    '''
    totalLen = len(inputArray)
    for i in range(totalLen):
        print("================================================")
        if analizeElement(inputArray[i], inputArray[:i] + inputArray[i + 1:], 1, totalLen):
            return True
        print("================================================")
    return False


def analizeElement(item, array, step, totalLen):
    print("Now analizing '{}' with array {} in step {}".format(item, array, step))
    if step == totalLen:
        return True
    for i in range(len(array)):
        if sum(1 for a, c in zip(item, array[i]) if a != c) == 1:
            if analizeElement(array[i], array[:i] + array[i + 1:], step + 1, totalLen):
                return True
    return False


print(stringsRearrangement(["abc",  "bef",  "bcc", "bec",  "bbc",  "bdc"]))
print(stringsRearrangement(["aba", "bbb", "bab"]))

File: arcade/go/arcade-intro/07_Through the Fog/033_stringsRearrangement.go
func stringsRearrangement(inputArray []string) bool {
	totalLen := len(inputArray)
	for i := range inputArray {
		if analizeElement(inputArray[i], getSliceWithoutNthElement(inputArray, i), 1, totalLen) {
			return true
		}
	}

	return false
}

func getSliceWithoutNthElement(a []string, n int) []string {
	cp := make([]string, len(a))
	copy(cp, a)
	b := append(cp[:n], cp[n+1:]...)
	return b
}

func analizeElement(item string, array []string, step int, totalLen int) bool {
	if step == totalLen {
		return true
	}

	for i := range array {
		count := 0
		for j := range item {
			if item[j] != array[i][j] {
				count++
			}
		}

		if count == 1 {
			if analizeElement(array[i], getSliceWithoutNthElement(array, i), step+1, totalLen) {
				return true
			}
		}
	}
	return false
}
